### ISSUE TITLE
"Object reference not set to an instance of an object" on Recursive Dtos

### ISSUE THREAD

**Issue by tzisopoulos** [agileobjects/AgileMapper#182]
I would like your help to identify if what I found is bug/limitation or am I missing something.

 I have the following two DTos  (attached as source and target)

[SourceDto.txt](https://github.com/agileobjects/AgileMapper/files/3982371/SourceDto.txt)
[TargetDto.txt](https://github.com/agileobjects/AgileMapper/files/3982372/TargetDto.txt)

I try to map them using the following command

`var testacc = (dynamic)_mapper.Map(source).ToANew<TargetDto>(cfg => cfg.Ignore(x => x.CreatedAt).And.Ignore(x => x.UpdatedAt));
`
and I get the following results:

1. If the source has a value for ParentAccount then everything works.

2. If no ParentAccount exists then I get "Object reference not set to an instance of an object." 

3. If I change the ParentAccount on targetDto to ParentAccount2 and map those two directly *with map from,to), then it works as expected.

I think that something confuse the mapper when we have recursively maps.

**Comment by tzisopoulos** [comment]
Also DateTimeOffset? to DateTimeOffset is not supported or does it need a specific configuration?
For example decimal? to decimal works perfect but DateTimeOffset? to DateTimeOffset throws 
`The operands for operator 'Convert' do not match the parameters of method 'op_Implicit`

**Comment by SteveWilkes** [comment]
Hi,

Thanks for reporting these two issues - I'll check them out in due course )

Cheers,

Steve

**Comment by SteveWilkes** [comment]
Hello!

Sorry for the delay, there - I'm not actually able to reproduce the `ParentAccount` issue - I've made a .NET Fiddle [here](https://dotnetfiddle.net/Dwrm0r) using the latest v1.6 preview - could you update it to make it throw the null reference exception?

I'm looking at `DateTimeOffsets` now and will report back :)

Happy new year!

Steve

**Comment by SteveWilkes** [comment]
`DateTimeOffset` mapping is now supported in the latest v1.6 branch code - release to follow :)

**Comment by tzisopoulos** [comment]
Hi, 

Thanks a lot of the fast response and the fix on the DateTimeOffset. 👍 

Regarding the other problem I tried your fiddle and indeed it worked. In my solution though still doesn't work but I found something weird.
If I call **GetPlantFor** prior of the mapping, then everything works great, If I skip this method then I get the NullRerferenceError. 
Does the GetPlanFor setup a plan inside the mapper or it just creates a projection and returns it. 

```
//Get the mapping for debugging purposes
string mapping = _mapper.GetPlanFor(account).ToANew<AccountDto>();

//Do the actual mapping
AccountDto accountDto = _mapper.Map(account).ToANew<AccountDto>();
```

I have attached the full StackTrace of the error and the map planning for your reference. If you need anything else let me know and I will provide it asap. 
Also I forgot to mention that I creating a new instance of the mapper on the application startup and I am using Autofac for DI.

[NullRefereneError.txt](https://github.com/agileobjects/AgileMapper/files/4029620/NullRefereneError.txt)

[Mapping plan.txt](https://github.com/agileobjects/AgileMapper/files/4029614/Mapping.plan.txt)

**Comment by tzisopoulos** [comment]
One more question. Mapping from a dynamic object is not supported?
If i try to map from a dynamic object to a dto using the ToANew I get the error 

`'object' does not contain a definition for 'ToANew'`

I have created the following a fiddle to check it
[Fiddle](https://dotnetfiddle.net/FDjIS4)

Thanks again :)

**Comment by SteveWilkes** [comment]
Hello!

Dynamic does work, but you have to use `ExpandoObject` rather than anonymous types - there's a StackOverflow question [here](https://stackoverflow.com/questions/9416095/dynamic-does-not-contain-a-definition-for-a-property-from-a-project-reference) which explains why. I think it's failing in .NET Fiddle for the same reason.

I've looked into it a bit and it also doesn't seem to work in .NET Core 1.*, but my tests work fine in .NET 4 and .NET Core 2+.

```csharp
dynamic source = new ExpandoObject();
source.Line1 = "Over here!";
source.Line2 = "Over there!";

var address = Mapper.Map(source).ToANew<Address>();

Assert.NotNull(address);
Assert.Equal("Over here!", address.Line1);
Assert.Equal("Over there!", address.Line2);
```

Cheers,

Steve

**Comment by SteveWilkes** [comment]
Hello again!

Regarding `GetPlanFor` - it just up-front caches the same plan which is / should be generated during a mapping, so it is odd that that fixes the issue...

If you could edit [the fiddle](https://dotnetfiddle.net/Dwrm0r) to include your full DTOs - with the `ISet`s and the other class definitions - maybe that would make it fall over. Otherwise if `GetPlanFor` fixes it, I guess all's well that ends well...?

Cheers,

Steve

**Comment by tzisopoulos** [comment]
I've done some investigation and found the following:

Mapper works great if it is consumed by the same application, but it doesn't work if its a part of a nuget package.
For example I am creating a internal nuget package (from no and on let's call it **package1**) that uses a single instance of the mapper (DI using autofac and registered as SingleInstance) and at some point it does the following
`result.Account = _mapper.Map(account).ToANew<AccountDto>();` 
which works great for all of my integration tests inside the same application.

but from another application if download **package1** (from internal nuget repo) and call the method that holds the above mapping) I get a System.security.verificationexception ( operation could destabilize the runtime. ).

Everything works great though from any application if I call (on the **package1** startup configuration) 
```
     private IMapper AgileMapperConfigurations()
    {
            var mapper = Mapper.CreateNew();
            ConfigureAccount(mapper);

            //Validate mappings
            mapper.ThrowNowIfAnyMappingPlanIsIncomplete();

            mapper.GetPlanFor<Account>().ToANew<AccountDto>();
            return mapper;
    }
```

 What I suppose is that the GetPlanFor create and compile a plan and this will be used from any mapping afterwards. So if any subproperty is null or something else it will not be considered on the runtime.

If I do not call the GetPlanFor I get the verification error because a property could be null and this could lead to some nasty error on the application.

I came to this conclusion after the below two points
1.  With the GetPlanFor everything works
2. I have attached two logs that show the Cached plans before the GetPlanFor and afterwards.
     The first one is empty and the second holds all the plan with a comment that parentAccount data are null.

I do not know if all these help you identify the issue (if any), but I have also attached the dtos and the logs for your convenience

[AgileMapper entities and logs.zip](https://github.com/agileobjects/AgileMapper/files/4085771/AgileMapper.entities.and.logs.zip)

**Comment by SteveWilkes** [comment]
Ahh - the plot thickens! Thanks very much for going to the trouble to figure all that out!

When AgileMapper creates its mapping `Func`s, they get housed in a dynamically-generated assembly - I'm guessing that generating that assembly when it's created by the AgileMapper assembly _via_ your internal NuGet package has different security implications than accessing it when  it's created by code from the consuming application itself. What version of .NET Framework / .NET Core is the application executing on?

On a hunch - d'you get the same issue if the consuming application (and / or your internal NuGet package) has the following in it?

```csharp
[assembly: System.Security.AllowPartiallyTrustedCallers]
```

This attribute by-passes certain security checks when accessing dynamically-generated assemblies. It may or may not help - if you could let me know that'd be great!

**Comment by SteveWilkes** [comment]
Just considering why it only happens with the _recursive_ relationship, and it has to be because recursive relationship mapping plans are created mid-mapping when a recursive relationship needs to be mapped - that means it's the dynamically-generated assembly containing the _original_ mapping plan creating a new dynamically-generated assembly containing the _recursive_ mapping plan - that must be what's violating the security policy.

**Comment by tzisopoulos** [comment]
I suspect that this could be the reason I get this exception (it makes sense), but unfortunately I am not able to test it this way because Autofac returns the error 
`Attempt by security transparent method X to access security critical method Y failed`
 if I add the aforementioned assembly attribute on my namespace.

I will try again with a completely different application to see how it will behave and will get back

Thanks a lot! 👍

**Comment by SteveWilkes** [comment]
`DateTimeOffset` is included in v1.6, which is now available [on NuGet](https://www.nuget.org/packages/AgileObjects.AgileMapper), and I think I'll add a note to the documentation somewhere regarding the `SecurityException` you found.

Thanks again for the feedback!
